core/src/main/java/org/apache/calcite/prepare/CalcitePrepareImpl.java

```diff
@@ -537,6 +540,11 @@ protected RelOptPlanner createPlanner(
       planner.addRule(AbstractMaterializedViewRule.INSTANCE_JOIN);
       planner.addRule(AbstractMaterializedViewRule.INSTANCE_PROJECT_AGGREGATE);
       planner.addRule(AbstractMaterializedViewRule.INSTANCE_AGGREGATE);
+    } else {
+      // See [CALCITE-2223]. Enabling the rules with materializations might result in infinite
+      // planning. E.g. org.apache.calcite.test.MaterializationTest.testJoinMaterialization10
+      planner.addRule(FilterSetOpTransposeRule.INSTANCE);
```

Comment:
@julianhyde , I'm not sure if those rules should be activated by default, however I think we can do that.
Of course `FilterSetOpTransposeRule` does not currently play well with materializations, so I include certain rules under `if (not materializations_enabled)`

Objections?